src/ir/auth_logic/ast_construction.cc

```diff
@@ -40,71 +40,152 @@ using auth_logic_cc_generator::AuthLogicLexer;
-          &context)) {  // PredicateContext
+    Principal left_principal,
+    AuthLogicParser::VerbphraseContext& verbphrase_context) {
+  if (auto* predphrase_context =
+          dynamic_cast<AuthLogicParser::PredphraseContext*>(
+              &verbphrase_context)) {  // PredicateContext
     return BaseFact(Attribute(
-        left_principal, ConstructPredicate(*CHECK_NOTNULL(ctx->predicate()))));
+        left_principal,
+        ConstructPredicate(*CHECK_NOTNULL(predphrase_context->predicate()))));
   }
-  auto& ctx =
+  auto& act_as_phrase_context =
       *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::ActsAsPhraseContext*>(
-          &context));  // ActsAsPhraseContext
+          &verbphrase_context));  // ActsAsPhraseContext
   return BaseFact(CanActAs(
-      left_principal, ConstructPrincipal(*CHECK_NOTNULL(ctx.principal()))));
+      left_principal,
+      ConstructPrincipal(*CHECK_NOTNULL(act_as_phrase_context.principal()))));
 }
 
-static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
-  if (auto* ctx = dynamic_cast<AuthLogicParser::PrinFactContext*>(&context)) {
+static BaseFact ConstructFlatFact(
+    AuthLogicParser::FlatFactContext& flat_fact_context) {
+  if (auto* prin_fact_context =
+          dynamic_cast<AuthLogicParser::PrinFactContext*>(&flat_fact_context)) {
     return ConstructVerbphrase(
-        ConstructPrincipal(*CHECK_NOTNULL(ctx->principal())),
-        *CHECK_NOTNULL(ctx->verbphrase()));
+        ConstructPrincipal(*CHECK_NOTNULL(prin_fact_context->principal())),
+        *CHECK_NOTNULL(prin_fact_context->verbphrase()));
   }
-  auto& ctx = *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::PredFactContext*>(
-      &context));  // predFactContext
-  return BaseFact(ConstructPredicate(*CHECK_NOTNULL(ctx.predicate())));
+  auto& pred_fact_context =
+      *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::PredFactContext*>(
+          &flat_fact_context));  // predFactContext
+  return BaseFact(
+      ConstructPredicate(*CHECK_NOTNULL(pred_fact_context.predicate())));
+}
+
+static BaseFact ConstructRvalue(
+    AuthLogicParser::RvalueContext& rvalue_context) {
+  if (auto* flat_fact_rvalue_context =
+          dynamic_cast<AuthLogicParser::FlatFactRvalueContext*>(
+              &rvalue_context)) {
+    return ConstructFlatFact(
+        *CHECK_NOTNULL(flat_fact_rvalue_context->flatFact()));
+  }
+  auto& binop_rvalue_context = *CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::BinopRvalueContext*>(&rvalue_context));
+  std::vector<std::string> args = utils::MapIter<std::string>(
+      binop_rvalue_context.pred_arg(),
+      [](auto* pred_arg) { return pred_arg->getText(); });
+  return BaseFact(datalog::Predicate(binop_rvalue_context.binop()->getText(),
+                                     std::move(args),
+                                     datalog::Sign::kPositive));
 }
 
 static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
```

Comment:
It looks like the idea is to populate this list through the recursive calls to `ConstructFact` inside this code, and then whenever a `Fact` is made from some larger piece of syntax like `SaysAssertions` or `Queries`, you pass in an empty list. This makes sense. It might be easier to read though if you either:
 - 1)  Add a comment above this explaining that this list is populated by recursion and that when this is called from outside it should be passed an empty list
- 2) Rename this function to `ConstructFactInner(std::forward_list<Principal> delegation_chain, AuthLogicParser::FactContext& fact_context)` and make a new function with the name and signature `ConstructFact(AuthLogicParser::FactContext& fact_context)`

Or ideally, do both.

Reply:
Added comments for ConstructFact